web/models/releases.ex

```diff
@@ -0,0 +1,136 @@
+    Ecto.Multi.new
+    |> create_package(package, user, meta)
+    |> create_release(package, checksum, meta)
+    |> audit_publish(audit_data)
+    |> publish_release(body)
+    |> Repo.transaction(timeout: @publish_timeout)
+    |> publish_result
+  end
+
+  def revert(package, release, [audit: audit_data]) do
+    delete_query =
+      from(p in Package,
+        where: p.id == ^package.id,
+        where: fragment("NOT EXISTS (SELECT id FROM releases WHERE package_id = ?)", ^package.id)
+      )
+
+    Ecto.Multi.new
+    |> Ecto.Multi.delete(:release, Release.delete(release))
+    |> audit_revert(audit_data, package, release)
+    |> Ecto.Multi.delete_all(:package, delete_query)
+    |> Repo.transaction_with_isolation(level: :serializable, timeout: @publish_timeout)
+    |> revert_result
+  end
+
+  defp publish_result({:ok, _} = result) do
+    HexWeb.RegistryBuilder.rebuild
+    result
+  end
+  defp publish_result(result), do: result
+
+  defp revert_result({:ok, %{release: release}}) do
+    revert_assets(release)
+    HexWeb.RegistryBuilder.rebuild
+    :ok
+  end
+  defp revert_result(result), do: result
+
+  defp create_package(multi, package, user, meta) do
+    params = %{"name" => meta["name"], "meta" => meta}
+    if package do
+      Ecto.Multi.update(multi, :package, Package.update(package, params))
+    else
+      Ecto.Multi.insert(multi, :package, Package.build(user, params))
+    end
+  end
+
+  defp create_release(multi, package, checksum, meta) do
+    version = meta["version"]
+    params = %{
+      "app" => meta["app"],
+      "version" => version,
+      "requirements" => normalize_requirements(meta["requirements"]),
+      "meta" => meta}
+
+    release = package && Repo.get_by(assoc(package, :releases), version: version)
+
+    if release do
+      release = Repo.preload(release, requirements: Release.requirements(release))
+      multi
+      |> Ecto.Multi.update(:release, Release.update(release, params, checksum))
+      |> Ecto.Multi.run(:action, fn _ -> {:ok, :update} end)
+    else
+      multi
+      |> build_release(params, checksum)
+      |> Ecto.Multi.run(:action, fn _ -> {:ok, :insert} end)
+    end
+  end
+
+  defp build_release(multi, params, checksum) do
+    Ecto.Multi.merge(multi, fn %{package: package} ->
+      Ecto.Multi.insert(Ecto.Multi.new, :release, Release.build(package, params, checksum))
+    end)
+  end
+
+  defp audit_publish(multi, audit_data) do
+    audit(multi, audit_data, "release.publish", fn %{package: pkg, release: rel} -> {pkg, rel} end)
+  end
+
+  defp audit_revert(multi, audit_data, package, release) do
+    audit(multi, audit_data, "release.revert", {package, release})
+  end
+
+  defp publish_release(multi, body) do
+    Ecto.Multi.run(multi, :assets, fn %{package: package, release: release} ->
+      push_assets(package, release.version, body)
+      {:ok, :ok}
+    end)
+  end
+
+  defp push_assets(package, version, body) do
```

Comment:
yeah, I wanted to propose something after merging this PR (and an upcoming one for DocsController). I haven't looked at the assets code across the app to closely yet, but perhaps `HexWeb.Assets` (that under the hood uses CDN and Store) would be good.